In [2]:
import numpy as numpy
import pandas as pd

In [3]:
from pathlib import Path

PROJECT_ROOT = Path.cwd().parent
DATA_DIR = PROJECT_ROOT / "data" / "raw"

print(DATA_DIR)

f:\Projects\Advanced-Predictive-Maintenance\data\raw


In [4]:
import pandas as pd

errors = pd.read_csv(DATA_DIR / "PdM_errors.csv")
telemetry = pd.read_csv(DATA_DIR / "PdM_telemetry.csv")
machines = pd.read_csv(DATA_DIR / "PdM_machines.csv")
maint = pd.read_csv(DATA_DIR / "PdM_maint.csv")
failures = pd.read_csv(DATA_DIR / "PdM_failures.csv")

In [5]:
errors.head()

,datetime,machineID,errorID
0,2015-01-03 07:00:00,1,error1
1,2015-01-03 20:00:00,1,error3
2,2015-01-04 06:00:00,1,error5
3,2015-01-10 15:00:00,1,error4
4,2015-01-22 10:00:00,1,error4


In [6]:
telemetry.head()

,datetime,machineID,volt,rotate,pressure,vibration
0,2015-01-01 06:00:00,1,176.217853,418.504078,113.077935,45.087686
1,2015-01-01 07:00:00,1,162.879223,402.747490,95.460525,43.413973
2,2015-01-01 08:00:00,1,170.989902,527.349825,75.237905,34.178847
3,2015-01-01 09:00:00,1,162.462833,346.149335,109.248561,41.122144
4,2015-01-01 10:00:00,1,157.610021,435.376873,111.886648,25.990511


In [7]:
machines.head()

,machineID,model,age
0,1,model3,18
1,2,model4,7
2,3,model3,8
3,4,model3,7
4,5,model3,2


In [8]:
maint.head()

,datetime,machineID,comp
0,2014-06-01 06:00:00,1,comp2
1,2014-07-16 06:00:00,1,comp4
2,2014-07-31 06:00:00,1,comp3
3,2014-12-13 06:00:00,1,comp1
4,2015-01-05 06:00:00,1,comp4


In [9]:
failures.head()

,datetime,machineID,failure
0,2015-01-05 06:00:00,1,comp4
1,2015-03-06 06:00:00,1,comp1
2,2015-04-20 06:00:00,1,comp2
3,2015-06-19 06:00:00,1,comp4
4,2015-09-02 06:00:00,1,comp4


In [10]:
print("Shape:", telemetry.shape)

telemetry.info()

Shape: (876100, 6)
<class 'pandas.DataFrame'>
RangeIndex: 876100 entries, 0 to 876099
Data columns (total 6 columns):
 #   Column     Non-Null Count   Dtype  
---  ------     --------------   -----  
 0   datetime   876100 non-null  str    
 1   machineID  876100 non-null  int64  
 2   volt       876100 non-null  float64
 3   rotate     876100 non-null  float64
 4   pressure   876100 non-null  float64
 5   vibration  876100 non-null  float64
dtypes: float64(4), int64(1), str(1)
memory usage: 56.0 MB


total 876099 rows and 6 columns and int and float and str data types

In [11]:
telemetry.isnull().sum()

datetime     0
machineID    0
volt         0
rotate       0
pressure     0
vibration    0
dtype: int64

Telemetry dataset is clean.

In [12]:
telemetry["machineID"].nunique()

100

In [14]:
telemetry["machineID"].min(),

(1,)

In [15]:
telemetry["machineID"].max()

100

No.of machines equal to 100
and range from 0 - 100

In [16]:
telemetry["datetime"] = pd.to_datetime(
    telemetry["datetime"]
)

In [17]:
print("Start Date:",
      telemetry["datetime"].min())

print("End Date:",
      telemetry["datetime"].max())

Start Date: 2015-01-01 06:00:00
End Date: 2016-01-01 06:00:00


The range is between Start Date: 2015-01-01 06:00:00
End Date: 2016-01-01 06:00:00


In [18]:
m1 = telemetry[
    telemetry["machineID"] == 1
]

In [ ]:
m1["datetime"].head(10)

we collected the readings of the machines of every one hour 
once

In [ ]:
telemetry.describe()

In [ ]:
import matplotlib.pyplot as plt

telemetry["rotate"].hist(
    bins=50,
    figsize=(8,4)
)

plt.title("rotate Distribution")
plt.show()

In [ ]:
import matplotlib.pyplot as plt

telemetry["volt"].hist(
    bins=50,
    figsize=(8,4)
)

plt.title("Voltage Distribution")
plt.show()

In [ ]:
import matplotlib.pyplot as plt

telemetry["pressure"].hist(
    bins=50,
    figsize=(8,4)
)

plt.title("pressure Distribution")
plt.show()

In [ ]:
import matplotlib.pyplot as plt

telemetry["vibration"].hist(
    bins=50,
    figsize=(8,4)
)

plt.title("vibration Distribution")
plt.show()

In [ ]:
import seaborn as sns

sns.boxplot(
    x=telemetry["vibration"]
)

In [ ]:
import seaborn as sns

sns.boxplot(
    x=telemetry["pressure"]
)

In [ ]:
import seaborn as sns

sns.boxplot(
    x=telemetry["volt"]
)

In [ ]:
import seaborn as sns

sns.boxplot(
    x=telemetry["rotate"]
)

In [31]:
failures[failures.machineID == 1]

,datetime,machineID,failure
0,2015-01-05 06:00:00,1,comp4
1,2015-03-06 06:00:00,1,comp1
2,2015-04-20 06:00:00,1,comp2
3,2015-06-19 06:00:00,1,comp4
4,2015-09-02 06:00:00,1,comp4
5,2015-10-17 06:00:00,1,comp2
6,2015-12-16 06:00:00,1,comp4


In [32]:
telemetry[
    (telemetry.machineID == 1)
]

,datetime,machineID,volt,rotate,pressure,vibration
0,2015-01-01 06:00:00,1,176.217853,418.504078,113.077935,45.087686
1,2015-01-01 07:00:00,1,162.879223,402.747490,95.460525,43.413973
2,2015-01-01 08:00:00,1,170.989902,527.349825,75.237905,34.178847
3,2015-01-01 09:00:00,1,162.462833,346.149335,109.248561,41.122144
4,2015-01-01 10:00:00,1,157.610021,435.376873,111.886648,25.990511
...,...,...,...,...,...,...
8756,2016-01-01 02:00:00,1,153.724780,450.254337,85.197443,43.698137
8757,2016-01-01 03:00:00,1,172.418099,375.701907,95.568265,42.111661
8758,2016-01-01 04:00:00,1,174.588145,460.505353,95.483913,33.293572
8759,2016-01-01 05:00:00,1,184.224228,438.823315,102.694358,46.609515


In [33]:
failures["failure"].value_counts()

failure
comp2    259
comp1    192
comp4    179
comp3    131
Name: count, dtype: int64

In [34]:
errors["errorID"].value_counts()

errorID
error1    1010
error2     988
error3     838
error4     727
error5     356
Name: count, dtype: int64

interesting Observation

Errors are much more common than failures.

3919 errors
761 failures

This is good.

Why?

Because errors may act as:

Early warning signals

before actual failure occurs.

Insight #1
comp2 is the most failure-prone component
34% of all failures

This means:

During feature analysis, pay special attention to sensor behavior before comp2 failures.